File: server/test/e2e/page-objects/TribeListPage.ts
import {browser, By, element} from "protractor";
import setLocation from "../setLocation";
import {TribeCardStyles, TribeListStyles} from "./Styles";

const newTribeButton = element(By.className(TribeListStyles.newTribeButton));
const tribeCardHeaderLocator = By.className(TribeCardStyles.header);

export default {
    waitForTribeListPage: function () {
        browser.wait(() => newTribeButton.isPresent(), 2000)
    },
    getTribeElements: function () {
        return element.all(By.className(TribeCardStyles.className));
    },
    getTribeNameLabel: function (tribeElement) {
        return tribeElement.element(tribeCardHeaderLocator);
    },
    getNewTribeButton: function () {
        return newTribeButton;
    },
    async goTo() {
        await setLocation('/tribes/');
        await browser.wait(() =>
            newTribeButton.isPresent(), 2000);
    }
}

File: e2e/src/e2eTest/kotlin/com/zegreatrob/coupling/e2e/test/TribeListPage.kt
package com.zegreatrob.coupling.e2e.test

import com.zegreatrob.coupling.model.party.PartyId
import com.zegreatrob.wrapper.wdio.WebdriverBrowser
import com.zegreatrob.wrapper.wdio.WebdriverElement

object TribeListPage : StyleSyntax {
    override val styles = loadStyles("tribe/TribeList")

    val newTribeButton by getting()

    val tribeCardElements get() = TribeCard.element().all()

    fun tribeCardElement(tribeId: PartyId) = WebdriverElement(
        ".${TribeCard.styles.className}[data-tribe-id=\"${tribeId.value}\"]"
    )

    suspend fun goTo() {
        WebdriverBrowser.setLocation("/tribes/")
        waitForPage()
    }

    suspend fun waitForPage() {
        WebdriverBrowser.waitUntil({ element().isPresent() }, 2000, "TribeListPage.waitForPage")
    }
}
